### ISSUE TITLE
Add articles about BML and manifests

### ISSUE THREAD

**Issue by Screwtapello** [byuu/byuu.net#1]
byuu.org had an article about manifests that mentioned BML at a high-level, but we could/should break that out in a bit more detail, so we can address the ambiguities that me and Alcaro discovered.

**Comment by Screwtapello** [comment]
When we've done this, we'll need to add links to byuu.org/preservation

**Comment by Screwtapello** [comment]
We should also have an article about splitting a SNES game with concatenated firmware into its component parts, as a reference for hypothetical emulator authors who might want to support that format.

Maybe even write some ISC-licensed example C code they can use.

**Comment by Screwtapello** [comment]
It would be helpful to convert an XML manifest to the BML format, if only to confirm I understand everything properly. It might also help other people figure out what's going on.

The one from the MMX3 Zero Project [might be a good candidate](https://github.com/bsnes-emu/bsnes/issues/17).

**Comment by n3vu0r** [comment]
Just in case you're interested in a BML format definition, I've written a [parser](https://github.com/qu1x/bml) by defining the [grammar](https://github.com/qu1x/bml/blob/master/src/bml.pest) using [PEG](https://pest.rs/book/grammars/peg.html).

**Comment by Screwtapello** [comment]
Huh. I had assumed that since BML used significant whitespace, it wouldn't fit formal parsing techniques like PEG or LALR or whatever. That's good to know!

It'd also be nice to link to some sample implementations, so a wider variety of implementation languages and styles would be good.

Unfortunately the forum archives are offline, so I can't dig out the list of corner-case I came up with when I worked on my own parser, maybe I can persuade Kawa to look for them. I should probably also publish the test-cases I came up with, too.

**Comment by n3vu0r** [comment]
You're right, it's not pure PEG, it uses a stack extension with `PUSH`/`PEEK_ALL`/`DROP` calls to support indentation-sensitive languages which might vary among PEG parsers.

There is [bml-fuzz](https://gitlab.com/higan/bml-fuzz) testing against:

```
Parent-1.0=A-value child child="B value" child:C:"value"
  child:D:"value"
    grandchild
  child grandchild=A
    :multi-line
    :text-field
    grandchild=B foo=bar
      foo=bar

Parent-1.0
```

**Comment by Screwtapello** [comment]
Here's some documentation byuu wrote on HN back in the day: https://news.ycombinator.com/item?id=8645591

**Comment by Screwtapello** [comment]
I think [this Ben Eater video](https://www.youtube.com/watch?v=yl8vPW5hydQ&list=PLowKtXNTBypFbtuVMUVXNR0z1mu7dp7eH&index=2) would be a great introduction to the basic idea of what a memory map is and why things are usually more complex than "byte X of the ROM dump is at address X in memory".
